src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs

```diff
@@ -229,13 +227,14 @@ public static class ResolverUtility
         /// <param name="remoteWalkContext">remote Providers (all sources, including file sources)</param>
         /// <param name="cancellationToken">cancellation token</param>
         /// <returns>The requested range and remote match.</returns>
-        public static Task<Tuple<LibraryRange, RemoteMatch>> FindPackageLibraryMatchCachedAsync(
-            ConcurrentDictionary<LibraryRange, Task<Tuple<LibraryRange, RemoteMatch>>> cache,
+        public static async Task<Tuple<LibraryRange, RemoteMatch>> FindPackageLibraryMatchCachedAsync(
             LibraryRange libraryRange,
-           RemoteWalkContext remoteWalkContext,
+            RemoteWalkContext remoteWalkContext,
             CancellationToken cancellationToken)
         {
-            return cache.GetOrAdd(libraryRange, (cacheKey) => ResolvePackageLibraryMatchAsync(libraryRange, remoteWalkContext, cancellationToken));
+            Tuple<LibraryRange, RemoteMatch> result = await remoteWalkContext.ResolvePackageLibraryMatchCache.GetOrAddAsync(libraryRange, () => ResolvePackageLibraryMatchAsync(libraryRange, remoteWalkContext, cancellationToken), cancellationToken);
```

Comment:
you just return the task here to avoid the extra state machine.

Reply:
Done